evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,214 @@
+import networkx as nx
+from networkx.algorithms.dag import topological_sort
+
+from evalml.exceptions import MissingComponentError
+from evalml.model_family import ModelFamily
+from evalml.pipelines.components import ComponentBase
+from evalml.pipelines.components.utils import handle_component_class
+from evalml.utils import import_or_raise
+
+
+class ComponentGraph:
+    def __init__(self, component_names=None, edges=None, random_state=0):
+        """ Initializes a component graph for a pipeline as a DAG.
+
+        Arguments:
+            component_names (dict): Of the form {name: component} pairs, where
+                             `name` is a unique string for that component and
+                             `component` is either a string component name
+                             as recognized by evalml or a direct evalml
+                             component class
+            edges (list): A list of tuples of the form (from_component,
+                          to_component), referring to the components by the
+                          names as represented in the `component_names` dict
+        """
+        self.component_names = component_names or {}
+        self._component_graph = nx.DiGraph()
+        if edges:
+            self._component_graph.add_edges_from(edges)
+        self._compute_order = topological_sort(self._component_graph)
+        self.random_state = random_state
+
+    def from_list(self, component_list):
+        """Constructs a linear graph from a given list
+
+        Arguments:
+            component_list (list): String names or ComponentBase subclasses in
+                                   an order that represents a valid linear graph
+        """
+        for idx, component in enumerate(component_list):
+            component_class = handle_component_class(component)
+            component_name = component_class.name
+
+            child = None
+            if idx != len(component_list) - 1:
+                child = [handle_component_class(component_list[idx + 1]).name]
+            self.add_node(component_name, component_class, children=child)
+        return self
+
+    def instantiate(self, parameters):
+        """Instantiates all components within the graph using the given parameters
+
+        Arguments:
+            parameters (dict): Dictionary with component names as keys and dictionary of that component's parameters as values.
+                               An empty dictionary {} implies using all default values for component parameters.
+        """
+        for component_name, component_class in self.component_names.items():
+            if isinstance(component_class, ComponentBase):
```

Comment:
Maybe we should continue here rather than throwing a ValueError? Right now, we let users define graphs with a mix classes and instances and by raising we don't let them instantiate the classes.

```python
cg = ComponentGraph() 
cg.add_node('OneHot', OneHotEncoder())
cg.add_node("OneHot2", OneHotEncoder)
cg.instantiate({"OneHot2": {'top_n': 5}})
```

Reply:
Oh, this is a really interesting point I overlooked! The original intention was to disallow constructing graphs with instances, and that way the graph is "static" until instantiated within the Pipeline class that contains it. However, I can see the added benefit of flexibility in allowing this mix of classes and instances, so I'll update this!